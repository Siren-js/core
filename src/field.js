import * as coerce from './util/coerce';
import extendWith from './util/extend-with';
import { isNonNullObject, isNullish, isString } from './util/type-guard';

/**
 * @typedef {object} FieldOptions Optional Field members and extensions
 * @property {string | string[]} [class] A list of strings describing aspects of
 *    the `Field` based on the current representation. Possible values are
 *    implementation-dependent and should be documented. Setting the value to a
 *    `string` will result in a singleton array.
 * @property {string} [title] Textual annotation of a field. Clients may use
 *    this as a label.
 * @property {string} [type] Input type of the field. May include any of the
 *    [input types from HTML](https://html.spec.whatwg.org/multipage/input.html#the-input-element).
 * @property {any} [value] The value assigned to the `Field`.
 */

/**
 * Represents a control inside an `Action`. Serialization of a `Field` depends
 * on its `type` and its corresponding `Action`'s `type`.
 */
export class Field {
  #name;
  #class;
  #title;
  #type;
  #value;

  /**
   * @param {string} name A name describing the control. Must be unique within
   *    an `Action`.
   * @param {FieldOptions} options Optional `Field` members (`class`, `title`,
   *    `type`, `value`) and extensions
   */
  constructor(name, options = {}) {
    if (!isString(name)) {
      throw new TypeError('Field.name must be a string');
    }

    const { class: fieldClass, title, type, value, ...extensions } =
      options ?? {};

    this.#name = name;
    this.class = fieldClass;
    this.title = title;
    this.type = type;
    this.value = value;

    extendWith(this, extensions);
  }

  /**
   * A name describing the control. Must be unique within an `Action`.
   * @type {string}
   */
  get name() {
    return this.#name;
  }

  /**
   * A list of strings describing aspects of the `Field` based on the current
   * representation. Possible values are implementation-dependent and should be
   * documented. Setting the value to a `string` will result in a singleton
   * array.
   * @type {readonly string[]}
   */
  get class() {
    return this.#class;
  }

  set class(value) {
    this.#class = coerce.toOptionalStringArray(value, this.class);
  }

  /**
   * Textual annotation of a field. Clients may use this as a label.
   * @type {string}
   */
  get title() {
    return this.#title;
  }

  set title(value) {
    this.#title = coerce.toOptionalString(value, this.title);
  }

  /**
   * Input type of the field. May include any of the
   * [input types from HTML](https://html.spec.whatwg.org/multipage/input.html#the-input-element).
   * When missing, the default is assumed to be `text`.
   * @type {string}
   */
  get type() {
    return this.#type;
  }

  set type(value) {
    this.#type = coerce.toOptionalString(value, this.type);
  }

  /**
   * The value assigned to the `Field`.
   * @type {any}
   */
  get value() {
    return this.#value;
  }

  set value(value) {
    this.#value = isNullish(value) ? undefined : value;
  }

  /**
   * Customizes JSON serialization (via `JSON.stringify()`) to include
   * properties defined as getters
   */
  toJSON() {
    const { name, class: fieldClass, title, type, value, ...extensions } = this;
    return { name, class: fieldClass, title, type, value, ...extensions };
  }

  /**
   * Determines whether `value` is a parsable Siren field (i.e., can be passed
   * to `Field.of`)
   * @param {unknown} value
   * @returns {boolean}
   */
  static isValid(value) {
    return (
      value instanceof Field || (isNonNullObject(value) && isString(value.name))
    );
  }

  /**
   * Constructs a `Field` instance from any object. Use `Field.isValid`
   * beforehand to avoid unexpected behavior.
   * @param {Record<string, unknown>} value
   * @returns {Field}
   */
  static of(value) {
    if (value instanceof Field) {
      return value;
    }
    const { name, ...rest } = value;
    return new Field(name, rest);
  }
}
