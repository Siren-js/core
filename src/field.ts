/**
 * Copyright (C) 2020  Dillon Redding
 *
 * See the full notice in the LICENSE file at the top level of the repository.
 */
import * as coerce from './coerce';
import deepFreeze from './deep-freeze';
import Extendable from './extendable';
import { isRecord, isString, isStringArray, isUndefined } from './type-guards';

/**
 * Creates a `ParsedField` object
 * @param name Name of the field
 * @param options Object containing optional field members (e.g., `title`, `type`) and extensions
 */
export function field<T>(name: string, options: OptionalFieldMembers<T> = {}): ParsedField<T> {
    const { 'class': fieldClass, title, type, value, ...extensions } = options;
    return deepFreeze({
        name: coerce.toString(name),
        class: coerce.toOptionalStringArray(fieldClass),
        title: coerce.toOptionalString(title),
        type: coerce.toOptionalString(type),
        value,
        ...extensions,

        update<U>(value: U): ParsedField<U> {
            return field(name, { ...options, value });
        }
    });
}

export type OptionalFieldMembers<T> = Extendable & Pick<ParsableField<T>, 'class' | 'title' | 'type' | 'value'>;

export interface ParsableField<T = unknown> extends Extendable {
    class?: string[] | string;
    name: string;
    title?: string;
    type?: string;
    value?: T;
}

export interface ParsedField<T = unknown> extends Field<T> {
    /**
     * Update the field's value.
     * @returns A new `Field` object with the updated value
     */
    update<U>(value: U): ParsedField<U>;
}

/**
 * Describes a Siren [field](https://github.com/kevinswiber/siren#fields-1),
 * which represents a control inside of an `Action`.
 */
export interface Field<T = unknown> extends Extendable {
    /**
     * Describes aspects of the field based on the current representation.
     */
    readonly class?: readonly string[];

    /**
     * A name describing the control.
     */
    readonly name: string;

    /**
     * Textual annotation of a field. Clients may use this as a label.
     */
    readonly title?: string;

    /**
     * The input type of the field. This may include
     * [input types](https://html.spec.whatwg.org/#the-input-element) specified
     * in HTML5 (see the `FieldType` enumeration). The default value is
     * `'text'`.
     */
    readonly type?: string;

    /**
     * A value assigned to the field.
     */
    readonly value?: Readonly<T>;
}

/**
 * Determines whether `value` is a valid Siren field. Note that this is a strict
 * check. To see if `value` is parsable by `field()`, use `isParsableField()`.
 */
export function isField(value: unknown): value is Field {
    return isParsableField(value) &&
        (isUndefined(value.class) || isStringArray(value.class)) &&
        (isUndefined(value.title) || isString(value.title)) &&
        (isUndefined(value.type) || isString(value.type));
}

/**
 * Determines whether `value` is a parsable Siren field. This function differs
 * from `isField()` in that the loose parsing rules of `field()` are taken into
 * consideration.
 */
export function isParsableField(value: unknown): value is ParsableField {
    return isRecord(value) && isString(value.name);
}

/**
 * Defines [input types](https://html.spec.whatwg.org/#attr-input-type-keywords)
 * specified in HTML5. `enum` names correspond to the input state and the values
 * correspond to the input keyword.
 */
export enum FieldType {
    Hidden = 'hidden',
    Text = 'text',
    Search = 'search',
    Telephone = 'tel',
    Url = 'url',
    Email = 'email',
    Password = 'password',
    Date = 'date',
    Month = 'month',
    Week = 'week',
    Time = 'time',
    LocalDateAndTime = 'datetime-local',
    Number = 'number',
    Range = 'range',
    Color = 'color',
    Checkbox = 'checkbox',
    RadioButton = 'radio',
    FileUpload = 'file',
    SubmitButton = 'submit',
    ImageButton = 'image',
    ResetButton = 'reset',
    Button = 'button'
}
