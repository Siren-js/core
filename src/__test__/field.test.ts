import * as Siren from '../field';

describe('field function', () => {
    it('should coerce name parameter to string', () => {
        const cases = [
            ['foo', 'foo'],
            ['foo bar', 'foo bar'],
            ['', ''],
            [undefined, ''],
            [null, ''],
            [true, 'true'],
            [42, '42'],
            [[true, 42, 'foo'], 'true,42,foo'],
            [{}, '[object Object]']
        ];

        cases.forEach(([value, expected]) => {
            const field = Siren.field(value as string);
            expect(field.name).toEqual(expected);
            field.name = value as string;
            expect(field.name).toEqual(expected);
        });
    });

    describe('optional parameter', () => {
        it('should set optional members to undefined when empty', () => {
            const name = 'orderNumber';

            const field = Siren.field(name);

            expect(field.name).toEqual(name);
            expect(field.class).toBeUndefined();
            expect(field.title).toBeUndefined();
            expect(field.type).toBeUndefined();
            expect(field.value).toBeUndefined();
        });

        it('should coerce class option to string array or undefined', () => {
            const cases = [
                [['order'], ['order']],
                [['customer', 'info'], ['customer', 'info']],
                [[], []],
                [undefined, undefined],
                [null, undefined],
                [true, ['true']],
                [42, ['42']],
                ['foo', ['foo']],
                [[true, 42, 'foo'], ['true', '42', 'foo']],
                [{}, ['[object Object]']]
            ];

            cases.forEach(([value, expected]) => {
                const field = Siren.field('name', { class: value as string[] });
                expect(field.class).toEqual(expected);
                field.class = value as string[];
                expect(field.class).toEqual(expected);
            });
        });

        it('should coerce title option to string or undefined', () => {
            const cases = [
                ['foo', 'foo'],
                ['foo bar', 'foo bar'],
                ['', ''],
                [undefined, undefined],
                [null, undefined],
                [true, 'true'],
                [42, '42'],
                [[true, 42, 'foo'], 'true,42,foo'],
                [{}, '[object Object]']
            ];

            cases.forEach(([value, expected]) => {
                const field = Siren.field('name', { title: value as string });
                expect(field.title).toEqual(expected);
                field.title = value as string;
                expect(field.title).toEqual(expected);
            });
        });

        it('should coerce type option to string or undefined', () => {
            const cases = [
                ['foo', 'foo'],
                ['foo bar', 'foo bar'],
                ['', ''],
                [undefined, undefined],
                [null, undefined],
                [true, 'true'],
                [42, '42'],
                [[true, 42, 'foo'], 'true,42,foo'],
                [{}, '[object Object]']
            ];

            cases.forEach(([value, expected]) => {
                const field = Siren.field('name', { type: value as string });
                expect(field.type).toEqual(expected);
                field.type = value as string;
                expect(field.type).toEqual(expected);
            });
        });

        it('should not modify value option', () => {
            [undefined, null, true, 42, 'foo', [true, 42, 'foo'], { a: 1, b: 2 }].forEach(value => {
                const field = Siren.field('name', { value });
                expect(field.value).toEqual(value);
                field.value = value;
                expect(field.value).toEqual(value);
            });
        });

        it('should accept extensions', () => {
            const min = 0;
            const max = 5;

            const field = Siren.field('orderNumber', { min });
            field.max = max;

            expect(field.min).toEqual(min);
            expect(field.max).toEqual(max);
        });

        it('should override required parameters', () => {
            const field = Siren.field('orderNumber', {
                name: 'quantity'
            });

            expect(field.name).toEqual('quantity');
        });
    });
});

describe('field type guard', () => {
    it('should return true for valid fields', () => {
        const values: Siren.Field[] = [
            // minimal
            { name: 'orderNumber' },
            // kitchen sink
            {
                name: 'orderNumber',
                class: ['integer'],
                title: 'Order Number',
                type: Siren.FieldType.Number,
                value: 42
            },
            // with extension
            { name: 'orderNumber', min: 0 }
        ];

        values.forEach(value => {
            expect(Siren.isField(value)).toEqual(true);
        });
    });

    it('should return false for invalid links', () => {
        const values = [
            // non-records
            undefined, null, true, 42, 'foo', [],
            // empty record
            {},
            // invalid name
            { name: true },
            // invalid class
            { name: 'orderNumber', class: 'order' },
            // invalid title
            { name: 'orderNumber', title: 42 },
            // invalid type
            { name: 'orderNumber', type: 42 }
        ];

        values.forEach(value => {
            expect(Siren.isField(value)).toEqual(false);
        });
    });
});

test('Field serialization', () => {
    const field = Siren.field('quantity', {
        type: 'number',
        title: 'Quantity',
        class: ['integer'],
        min: 1,
        max: 10
    });

    const json = JSON.stringify(field, null, 2);

    expect(json).toMatchSnapshot();
});
