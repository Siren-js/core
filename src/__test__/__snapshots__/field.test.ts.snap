// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field serialization 1`] = `
"{
  \\"name\\": \\"quantity\\",
  \\"class\\": [
    \\"integer\\"
  ],
  \\"title\\": \\"Quantity\\",
  \\"type\\": \\"number\\",
  \\"min\\": 1,
  \\"max\\": 10
}"
`;
